### ISSUE TITLE
数据库连接池应用

### ISSUE THREAD

**Issue by kycool** [defrur/tech-notes#20]
mysql 数据库连接池的应用

> 之前很少应用到 `mysql 连接池`，现在也是相当于学习

- `mysql 数据库连接池` 结合 `Python` 的具体实例和测试
- `mysql 数据库连接池` 在 `django` 中的应用和分析

**Comment by defrur** [comment]
对 MySQL 来说，`连接池` 是客户端的概念 （以下客户端特指数据库客户端，也就是 `mysqld` 对应的客户端）。放到 `Django` 环境下，客户端也就是 `Django` 。通常说的连接池是指客户端维护一个池子，里面放有若干个从客户端到 `mysqld` 的数据库链接。以便减少客户端和 `mysqld` 建立连接时的开销。

但是官方文档看，`Django`  并没有实现连接池，而是只实现了持久连接。相当于只实现了大小为 `1 ` 的连接池。

默认情况下，`Django` 并不开启持久连接。控制该行为的是  [CONN_MAX_AGE](https://docs.djangoproject.com/en/1.8/ref/settings/#std:setting-CONN_MAX_AGE) 。

>The default value is 0, preserving the historical behavior of closing the database connection at the end of each request. 

每次请求的时候都会打开一个连接，请求结束时关闭连接。

当 `CONN_MAX_AGE` 的值大于 `0` 时，Django 才会开启持久连接。


## 如何使用连接池？

使用第三方的 ORM 方案即可开启连接池。例如 `SQLAlchemy`。

## 使用连接池有什么后果？
从默认配置下得知 `Django` 并不推荐使用持久连接，更不用说连接池了。从我的实践经验看，开启了数据库连接池意味着多了很多麻烦事。除非你知道你在做什么，否则不要轻易开启连接池。

可能造成的后果有：

#### 连接池不够用

开启了连接池之后，数据库连接通常是和 `Request` 有关，`Request` 数量增大的时候，连接池就不够用了。例如你连接池的大小设置为 `30` ，每个 `Request` 都会使用一个数据库连接。当并发数 > `30` 的时候，应用就无法从连接池获得一个数据库连接。

也许你会尝试重新连接池的大小，但是连接池的大小还受到另外一个制约。稍有不当，会引起另外一个问题：数据库服务端连接数过高。

#### 数据库服务端连接数过高

每一个数据库链接在 `mysqld` 中都会占用的资源，即使是一个空的数据库链接也会引起性能损耗。`mysql` 的连接数是由 [max_connections](https://dev.mysql.com/doc/refman/5.5/en/server-system-variables.html#sysvar_max_connections) 参数控制的。默认值是 `151`。超过这个数的时候，客户端是不能连接到 `mysqld` 的。而这个 `max_connections` 直接和数据库主机的硬件有关系，显然不能设置成无限大。


## 非要使用怎么办？

1. 首先要给出充分的理由来使用连接池，不能想当然，要有数据支撑。
2. 小心地设置连接池的大小。注意，如果你部署了N 个进程。那么产生的连接数就是 N * size_of_pool。
3. 优化好代码，每个 Requests 的响应时间尽可能短，否则尽早 `显式` 地关闭数据库连接。
4. 配置更好的主机来运行 MySQL。
5. 做好监控。
